src/client.rs

```diff
@@ -0,0 +1,162 @@
+        let nb_leaves = (nb_dummies + 1) / 2;
+        let mut dummies = Vec::new();
+
+        let cipher = Aes256Gcm::new(&self.key);
+
+        // FIXME - encrypt fixed dummy value instead of encrypting randoms.
+        for _ in 0..nb_dummies * BUCKET_SIZE {
+            // Generate new random dummy data.
+            let mut dummy_data = vec![0; block_size];
+            self.csprng.fill_bytes(&mut dummy_data);
+
+            // Generate new nonce for encryption.
+            let nonce = Nonce::new(&mut self.csprng);
+
+            // Encrypt dummies to provide correct MAC for later decryption.
+            let ciphertext_res =
+                cipher.encrypt(&nonce, &dummy_data, Option::None);
+
+            if ciphertext_res.is_err() {
+                return Err(Error::new(
+                    ErrorKind::Interrupted,
+                    Result::unwrap_err(ciphertext_res).to_string(),
+                ));
+            }
+
+            let encrypted_dummy =
+                [nonce.as_bytes(), ciphertext_res.unwrap().as_slice()].concat();
+
+            // FIXME- uniform generation for now. Is fine but dummies' paths do
+            // not need to be generated at random.
+            let path = self.csprng.gen_range(0..nb_leaves);
+
+            dummies.push(DataItem::new(encrypted_dummy, path as u16));
+        }
+
+        Ok(dummies)
+    }
+
+    pub fn encrypt_blocks(
+        &mut self,
+        blocks: &mut Vec<DataItem>,
+        block_ids: Vec<usize>,
+        max_path: usize,
+    ) -> Result<(), CryptoCoreError> {
+        self.gen_key();
+        let cipher = Aes256Gcm::new(&self.key);
+
+        let mut i = 0;
+        while i < blocks.len() {
+            // Generate new nonce for encryption.
+            let nonce = Nonce::new(&mut self.csprng);
+
+            let ciphertext_res =
+                cipher.encrypt(&nonce, blocks[i].data(), Option::None);
+            if ciphertext_res.is_err() {
+                return Err(Result::unwrap_err(ciphertext_res));
+            }
+
+            // Change element data to plaintext.
+            blocks[i].set_data(
+                [nonce.as_bytes(), ciphertext_res.unwrap().as_slice()].concat(),
+            );
+
+            /*
+             * If the block is among the ones to change, change its path by
+             * sampling a random uniform distribution.
+             */
+            if block_ids.contains(&i) {
+                blocks[i].set_path(self.csprng.gen_range(0..max_path as u16))
+            }
+
+            i += 1;
+        }
+
+        Ok(())
+    }
+
+    pub fn decrypt_blocks(
+        &self,
+        blocks: &mut Vec<DataItem>,
+    ) -> Result<(), CryptoCoreError> {
+        let cipher = Aes256Gcm::new(&self.key);
+
+        let mut i = 0;
+        while i < blocks.len() {
+            // Edge-case where dummies left cells uninitialized.
+            if blocks[i].data().is_empty() {
+                i += 1;
+                continue;
+            }
```

Comment:
Loop on an iterator instead.
```suggestion
        for block in blocks {
            // Edge-case where dummies left cells uninitialized.
            if block.data().is_empty() {
                continue;
            }
```

Reply:
But doesn't this method make a copy of the element we are working with ?
If I change block, do I change a copy of it or is it passed by reference ?

If the latter, then this method sounds better.

Reply:
fixed for an enumerator in commit 6b498d41c1a8a025a0d7f2e6f7d04a456619a1b8